src/core/lib/security/credentials/google_default/compute_engine_channel_credentials.cc

```diff
@@ -0,0 +1,70 @@
+/*
+ *
+ * Copyright 2020 The gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/security/credentials/credentials.h"
+
+#include <string.h>
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/sync.h>
+
+#include "src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.h"
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/gpr/env.h"
+#include "src/core/lib/gpr/string.h"
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+#include "src/core/lib/http/httpcli.h"
+#include "src/core/lib/http/parser.h"
+#include "src/core/lib/iomgr/load_file.h"
+#include "src/core/lib/iomgr/polling_entity.h"
+#include "src/core/lib/security/credentials/alts/alts_credentials.h"
+#include "src/core/lib/security/credentials/alts/check_gcp_environment.h"
+#include "src/core/lib/security/credentials/google_default/google_default_credentials.h"
+#include "src/core/lib/security/credentials/jwt/jwt_credentials.h"
+#include "src/core/lib/security/credentials/oauth2/oauth2_credentials.h"
+#include "src/core/lib/slice/slice_internal.h"
+#include "src/core/lib/slice/slice_string_helpers.h"
+#include "src/core/lib/surface/api_trace.h"
+
+grpc_channel_credentials* grpc_compute_engine_channel_credentials_create(
```

Comment:
This looks like a simplified version of `grpc_google_default_credentials_create()`.  Why can't that just be used as-is?  Why is the simplified version necessary?

Reply:
This is based on a proposal that @apolycn put together at go/yoshi-directpath. He can correct anything I get wrong. The problem is that `google_default_credentials` create their *own* call creds based on the default service account of the VM. However, the auth libraries for wrapped languages, being **dependents** of gRPC and not dependencies, may actually be asserting a different identity than the default service account of the VM.

When this mismatch occurs and Directpath is not enabled, traffic will flow just fine. When it's enabled in the console, however, RPCs will start to fail.

Unfortunately, `grpc_google_default_credentials_create` didn't add a `void* reserved` argument, so we can't just extend that function. You'll notice I tried to reuse as much of the existing implementation as possible. You can more or less think of this as `grpc_google_default_credentials_create2` with naming aligned to the other languages.

With that said, I'm more than open to alternate solutions to that problem.

Reply:
As @gnossen mentioned, using compute engine default engine is the way we agreed with the client library to address the identity mismatch issue (ALTS only uses the VM service account as identity, but GDC may use identity from environment or file which causes identity mismatch). Java and Go both have Compute Engine Creds/Channels.

Reply:
> Unfortunately, `grpc_google_default_credentials_create` didn't add a `void* reserved` argument, so we can't just extend that function.

We have no requirement to maintain backward compatibility in the C-core API.  The only requirement is that we publish a gRFC describing those changes.

If this is essentially just a way of overriding the call creds attached to GoogleDefaultCreds, then let's just add an optional parameter to the existing function that allows specifying custom call creds.